File: cpp/sort/mergeSort.cpp
#include <iostream>
#include <span>

using namespace std;


	/*
		Merge Sort
	*/

int* interlace(span<int> lh, span<int> rh)
{
	const size_t N = lh.size() + rh.size();
	int* laced = new int[N];

	int index = 0, lhi = 0, rhi = 0;

	while (index < N)
	{
		if (lhi > lh.size() - 1)
			laced[index] = rh[rhi++];
		else if (rhi > rh.size() - 1)
			laced[index] = lh[lhi++];
		else
			laced[index] = (lh[lhi] < rh[rhi] ? lh[lhi++] : rh[rhi++]);
		++index;
	}
	return laced;
}

void mergeSort(span<int> window)
{
	if (window.size() == 1)
		return;
	size_t mp = window.size() / 2;
	span<int> lh = window.subspan(0, mp);
	span<int> rh = window.subspan(mp);
	mergeSort(lh);//sort left half (including midpoint)
	mergeSort(rh);//sort right half starting one after the midpoint
	//merge halves
	int* sorted = interlace(lh, rh);
	for (int x{}; x < window.size(); ++x)
		window[x] = sorted[x];
	delete[] sorted;
	return;
}

int main()
{
	const uint32_t N = 9;
	int a[]{ 8,2,5,7,6,1,0,3,4 };

	mergeSort(a);

	span<int> print(a);
	for (auto it = print.begin(); it != print.end(); ++it)
		cout << " " << *it;
	cout << endl;

	cin.get();
	return 0;
}

File: julia/mergeSort.jl
arr = [2, 1, 3, 7, 6, 4, 9, 11]

function merge!(arr, l1, r1, l2, r2)
    merged = []
    start = l1
    while l1 <= r1 || l2 <= r2 
        if l1 > r1
            push!(merged, arr[l2])
            l2 += 1
        elseif l2 > r2 
            push!(merged, arr[l1])
            l1 += 1 
        elseif arr[l1] < arr[l2]
            l1 += 1
        else
            push!(merged, arr[l2])
            l2 += 1
        end
    end
    for i in eachindex(merged)
        arr[start] = merged[i]
        start += 1
    end
end

function mergesort!(arr, l, r)
    # base case
    r <= l && return

    mp = div(l+r, 2) 
    # 1 2 3 4
    # mergesort left half 
    mergesort!(arr, l, mp) # include midpoint in left half (arbitrary)
    # mergesort right half 
    mergesort!(arr, mp+1, r)

    # now merge the two halves
    merge!(arr, l, mp, mp+1, r)

    arr
end

n = length(arr)
mergesort!(arr, 1, n) |> println
